hadoop-hdfs-project/hadoop-hdfs-rbf/src/main/java/org/apache/hadoop/hdfs/server/federation/router/RouterRpcServer.java

```diff
@@ -361,6 +380,23 @@ public RouterRpcServer(Configuration configuration, Router router,
     this.nnProto = new RouterNamenodeProtocol(this);
     this.clientProto = new RouterClientProtocol(conf, this);
     this.routerProto = new RouterUserProtocol(this);
+
+    long dnCacheExpire = conf.getTimeDuration(
+        DN_REPORT_CACHE_EXPIRE,
+        DN_REPORT_CACHE_EXPIRE_MS_DEFAULT, TimeUnit.MILLISECONDS);
+    this.dnCache = CacheBuilder.newBuilder()
+        .build(new DatanodeReportCacheLoader());
+
+    // Actively refresh the dn cache in a configured interval
+    Executors
```

Comment:
Hmm, have you considered using
```java
this.dnCache = CacheBuilder.newBuilder()
         .refreshAfterWrite(dnCacheExpire, TimeUnit.MILLISECONDS)
         .build(new DatanodeReportCacheLoader());
```

This will also automatically refresh the caches. Also it only refreshes a key iff 1) it becomes stale, and 2) there is a request on it. So this will save some calls for those infrequent DN report types.

Reply:
Yes. The point here is, with refreshAfterWrite, you will only get the previously value in this call, but the result will be refreshed in the background for next retreival. If we only have 1 request per hour, you will only get the datanode report 1 hour ago, unless you make the call sync, which is slow. Given it is already a background thread and not that heavy with an interval, current design is better.

Reply:
I see. Makes sense.